Feature configurable grant types

angelakis opened the pull request:
As discussed in https://github.com/IdentityPython/oidcendpoint/pull/59#issuecomment-693260033, I split up the configurable grant types feature (originally by-grant-types) from the token exchange specific changes to make review easier.

This branch refactors token_coop module to allow configuring different handlers for each grant type configured.

rohe committed:
Refactored module.

angelakis committed:
Change how supported grant types are configured

The user can now set the usual grant types to have a default
configuration and class, without the need to set the oidcendpoint
class path. This can be done by setting a value of "default" or
None (no value in yaml dict).

angelakis committed:
Make multiple fixes

angelakis committed:
Add tests for configurable grant types

rohe commented:
EndpointHelper is not a general class for all endpoints so I think we should name it TokenEndpointHelper .

nsklikas commented on src/oidcendpoint/oidc/token_coop.py:
```diff
@@ -22,74 +24,93 @@
-    default_capabilities = {"token_endpoint_auth_signing_alg_values_supported": None}
+def aud_and_scope(url):
+    p = urlparse(url)
+    return "{}://{}".format(p.scheme, p.netloc), p.path
```
Remove this function

angelakis replied:
ok

nsklikas commented on src/oidcendpoint/oidc/token_coop.py:
```diff
@@ -22,74 +24,93 @@
-        _context = self.endpoint_context
+        request = AccessTokenRequest(**request.to_dict())
+
+        if "state" in request:
+            try:
+                sinfo = self.endpoint.endpoint_context.sdb[request["code"]]
+            except KeyError:
+                logger.error("Code not present in SessionDB")
+                return self.endpoint.error_cls(error="unauthorized_client")
+            else:
+                state = sinfo["authn_req"]["state"]
+
+            if state != request["state"]:
+                logger.error("State value mismatch")
+                return self.endpoint.error_cls(error="unauthorized_client")
```
Move this to separate function (`validate_state`)?

angelakis replied:
ok

nsklikas commented on src/oidcendpoint/oidc/token_coop.py:
```diff
@@ -131,82 +152,121 @@ def _access_token(self, req, **kwargs):
+        # TODO: do we want to allow any grant_type?
+        for grant_type, grant_type_options in grant_types_supported.items():
+            if (
+                    grant_type_options in ("default", None)
```
Add `True` as a default value?

angelakis replied:
Added True as a default value and removed string "default". To be more consistent, I also added False, which just ignores that particular grant type and doesn't enable it.

As a result, we can now use None or True to use the default grant type class. This may seem a bit counter-intuitive, but IMO makes sense, as our configs are YAML and I believe the following configuration options should be allowed for a better UX and should mean the same thing:
```
grant_types_supported:
  - authorization_code

grant_types_supported:
  - authorization_code: True
```

nsklikas commented on src/oidcendpoint/oidc/token_coop.py:
```diff
@@ -131,82 +152,121 @@ def _access_token(self, req, **kwargs):
+            except KeyError:
+                raise ProcessError(
+                        "Grant type is invalid or missing a valid class to import."
                 )
```
Improve error message?

Check if the required fields have been initialized and then initiate the class

angelakis replied:
OK I broke this part in multiple checks and exception handers.

nsklikas commented on tests/test_35_oidc_token_coop_endpoint.py:
```diff
@@ -158,6 +152,62 @@ def create_endpoint(self, conf):
+        token_conf = conf["endpoint"]["token"]
+        token_conf["kwargs"]["grant_types_supported"] = grant_types_supported
+        with pytest.raises(Exception):
+            EndpointContext(conf)
```
Check exception message

angelakis replied:
Did that (and also improved the exception handling in regards to that).

nsklikas commented on tests/test_35_oidc_token_coop_endpoint.py:
```diff
@@ -158,6 +152,62 @@ def create_endpoint(self, conf):
+        token_conf = conf["endpoint"]["token"]
+        token_conf["kwargs"]["grant_types_supported"] = grant_types_supported
+        endpoint_context = EndpointContext(conf)
+        assert endpoint_context
```
Add test for a custom grant?

angelakis replied:
Added a custom grant test.

angelakis committed:
Clean up things in token_coop and minor changes

Removed "default" value for grant types and added True instead.

So now None or True enables the grant_type with the default class,
while False disables it and of course a dict with a custom class
uses that instead.

angelakis commented:
> EndpointHelper is not a general class for all endpoints so I think we should name it TokenEndpointHelper .

Correct, I changed it.

angelakis commented:
Should I also change token_coop module and class name to Token and replace the existing one? Or leave it for the future?

rohe commented:
I think you should.

angelakis commented:
> I think you should.

Previously the token endpoint was named AccessToken instead of Token, which is the correct name now. I guess we should try to keep AccessToken (maybe as an alias) for backwards compatibility?

rohe commented:
I don't know. It is the token endpoint not the access token endpoint.
But sure keep AccessToken as an alias.

peppelinux reviewed (commented):
LGTM
@angelakis is it time to merge it?

angelakis commented:
> LGTM
> @angelakis is it time to merge it?

I wanted to rename to Token and create an alias for AccessToken but haven't found the time yet.

We can merge once I make the changes @nsklikas pointed today, the renaming can be done on a separate pr.

angelakis committed:
Improve exception handling in TokenCoop

angelakis committed:
Improve TokenCoop tests and add more

angelakis commented:
I believe this can be merged now.

peppelinux commented:
@rohe do we have to merge the new storage model before merge this or we can procede as It Is?

peppelinux commented:
This looks ready for merge. My fingers are on the merge Button, Is there any doubts? Can we proceed?

rohe commented:
Go ahead!
